src/charm.py

```diff
@@ -131,6 +131,10 @@ def _training_operator_layer(self) -> Layer:
 
     def _check_container_connection(self):
         """Check if connection can be made with container."""
+        # FIXME: if the container is unreachable, setting the unit
+        # to MaintenanceStatus won't have any effect in the execution
+        # which may cause issues if there is an attempt of starting
+        # the workload service or updating the pebble layer (which replans)
```

Comment:
but will not we check container connect each time when we call for container actions ?

we should check container can connect before each pebble layer update

Reply:
yes, what I mean in my comment is that there is no effect. We call the check method, if it does not connect, it sets the unit to maintenance status, and then proceeds with the execution. In the juju sdk it is recommended we defer this event and return, which we did in the past, to ensure that anything that depends on the container being reachable was able to run.

I did not add the defer here because deferring may require us to re-design the charm code.

Reply:
> basically, you need to defer only when you need to preserve order.

I disagree, deferring an event will prevent executing the next line after calling `defer()`, which can save us from attempting to do something in the workload container that will certainly fail.

I did not say defer is the only solution, though, my `FIXME` note just says that putting the unit in Waiting or MaintenanceStatus if the workload container cannot be reached won't have any effect and it will not stop the execution, which may cause errors.

Reply:
if you only want to exit, you can use sys.exit(0), that is fine

but defer on top of it preserves the order of the events. Considering following queue
eventA
eventB
eventC

then deferring in A, will try to always execute A before B and C. In case when you have know that the charm is not functionable until A is executed

Reply:
oh, but then we need either do:
```python
def foo():
  if container.can_connect():
    container.something_that_depends_on_container()
  else:
    model.unit.status = MaintenanceStatus
```

or if your code is deeply nested, eg, you check for connection in another function:
```python
def can_connect():
  if not container.can_connect():
    model.unit.status = MaintenanceStatus
    sys.exit(0)
  return True
```

Reply:
and in our case, we are raising, so, execution is broken, no?